ir/src/boundary_stmts.rs

```diff
@@ -0,0 +1,175 @@
+    /// last row.
+    pub fn last_boundary_constraints(
+        &self,
+        trace_segment: TraceSegment,
+    ) -> Vec<(usize, &BoundaryExpr)> {
+        if self.boundary_constraints.len() <= trace_segment.into() {
+            Vec::new()
+        } else {
+            self.boundary_constraints[trace_segment as usize]
+                .1
+                .iter()
+                .map(|(k, v)| (*k, v))
+                .collect()
+        }
+    }
+
+    // --- MUTATORS -------------------------------------------------------------------------------
+
+    /// Add a boundary statement from the AST. The statement can be either be a variable or a
+    /// constraint. In case it is a constraint, it is added to the boundary_constraints vector
+    /// in the relevant trace segment. Variables are currently not supported for boundary
+    /// constraints.
+    ///
+    /// # Errors
+    /// Returns an error if:
+    /// - The identifier specified for the boundary constraint column has not been declared or has
+    ///   been declared with the wrong type.
+    /// - The constraint expression is contains invalid public input references.
+    /// - A boundary constraint has already been declared for the specified column and boundary.
+    pub(super) fn insert(
+        &mut self,
+        symbol_table: &mut SymbolTable,
+        stmt: &BoundaryStmt,
+    ) -> Result<(), SemanticError> {
+        match stmt {
+            BoundaryStmt::Variable(_) => unimplemented!(),
+            BoundaryStmt::Constraint(constraint) => {
+                // validate the expression
+                let expr = constraint.value();
+                validate_expression(symbol_table, &expr)?;
+
+                // add the constraint to the specified boundary for the specified trace
+                let col_type = symbol_table.get_type(constraint.column())?;
+                let result = match col_type {
+                    IdentifierType::TraceColumn(column) => match column.trace_segment() {
+                        0 => match constraint.boundary() {
+                            ast::Boundary::First => self.boundary_constraints[0]
+                                .0
+                                .insert(column.col_idx(), expr),
+                            ast::Boundary::Last => self.boundary_constraints[0]
+                                .1
+                                .insert(column.col_idx(), expr),
+                        },
+                        1 => match constraint.boundary() {
+                            ast::Boundary::First => self.boundary_constraints[1]
+                                .0
+                                .insert(column.col_idx(), expr),
+                            ast::Boundary::Last => self.boundary_constraints[1]
+                                .1
+                                .insert(column.col_idx(), expr),
+                        },
+                        _ => unimplemented!(),
```

Comment:
With the new refactor making everything more generic we should be able to reduce redundancy and do something more elegant here.

Reply:
Thanks, please let me know if you had something else in mind.